### PR TITLE
Moved y-axis to right of secondary charts

### LINKED ISSUE #200: Add y-axis to MACD panel

### LINKED ISSUE #241: Y axis on right side of RSI
Y axis needs to be on the right side of the RSI chart for blog post.
(requested by Rob)

### PR DESCRIPTION
**Pull request by djmiley**
Resolves #200 . Resolves #241 

Re-implemented secondary charts using `linearTimeSeries`

### CODE REVIEW

File: src/assets/js/chart/macdChart.js [hunk 1]
```diff
@@ -2,11 +2,29 @@
...
+            .yValue(function(d, i) { return d.macd.macd; });
+        var signal = fc.series.line()
+            .yValue(function(d, i) { return d.macd.signal; });
+        var divergence = fc.series.bar()
+            .yValue(function(d, i) { return d.macd.divergence; });
 
-        var macd = fc.indicator.renderer.macd();
+        var multi = fc.series.multi()
+            .series([zero, macd, signal, divergence]);
+
```
Comment (jleft) [inline src/assets/js/chart/macdChart.js]: Can we use `fc.indicator.renderer.macd()`?

File: src/assets/js/chart/macdChart.js [hunk 2]
```diff
@@ -14,33 +32,48 @@
...
-                .range([0, fc.util.innerDimensions(selection.node()).width]);
-            macd.yScale()
-                .domain([-maxYExtent, maxYExtent])
-                .range([fc.util.innerDimensions(selection.node()).height, 0]);
+            // Redraw
+            macdTimeSeries.plotArea(multi);
+            selection.call(macdTimeSeries);
 
+            // Color axis lines so y-axis is emphasised
+            selection.select('g.y-axis').selectAll('path').style('stroke', '#000');
```
Comment (jleft) [inline src/assets/js/chart/macdChart.js]: Can this be done in Less?

File: src/assets/js/chart/rsiChart.js [hunk 3]
```diff
@@ -2,42 +2,70 @@
...
+        var annotations = fc.annotation.line()
+            .label('');
+        var indicator = fc.series.line()
+            .yValue(function(d, i) { return d.rsi; });
 
-        var rsi = fc.indicator.renderer.relativeStrengthIndex()
-            .yScale(rsiScale);
+        var multi = fc.series.multi()
+            .series([annotations, indicator]);
+
```
Comment (jleft) [inline src/assets/js/chart/rsiChart.js]: Can we use `fc.indicator.renderer.relativeStrengthIndex()`?

File: src/assets/js/chart/rsiChart.js [hunk 4]
```diff
@@ -2,42 +2,70 @@
...
+                            return 'multi ' + ['annotations', 'indicator'][i];
+                        });
+                });
+
+            // Redraw
+            rsiTimeSeries.plotArea(multi);
+            selection.call(rsiTimeSeries);
+
+            // Color y-axis so that it is not visible
+            selection.select('g.y-axis').selectAll('path').style('stroke', '#fefefe');
```
Comment (jleft) [inline src/assets/js/chart/rsiChart.js]: Can this be done in CSS?

File: src/assets/styles/variables.less [hunk 5]
```diff
@@ -2,6 +2,7 @@
 @primary-green: #6c0;
 @primary-red: #c60;
 @primary-white: #f0f8ff;
+@d3fc-offwhite: #fefefe;
```
Comment (jleft) [inline src/assets/styles/variables.less]: `@d3fc-offwhite` -> `@off-white`

File: src/assets/styles/primary.less [hunk 6]
```diff
@@ -0,0 +1,41 @@
+.primary {
+  .plot-area {
```
Comment (jleft) [inline src/assets/styles/primary.less]: As linear time series will introduce a plot area to the secondary charts, should these styles also apply to the secondary chart too to be consistent?

File: src/assets/js/chart/macdChart.js (hunk 1 above)
Comment (djmiley) [inline src/assets/js/chart/macdChart.js]: Need to look at this more

File: src/assets/js/chart/rsiChart.js (hunk 3 above)
Comment (djmiley) [inline src/assets/js/chart/rsiChart.js]: Need to look at this more

Comment (djmiley) [inline src/assets/js/chart/rsiChart.js]: I think this would be uglier.

File: src/assets/js/chart/macdChart.js (hunk 1 above)
Comment (djmiley) [inline src/assets/js/chart/macdChart.js]: I think this would be uglier

File: src/assets/js/chart/rsiChart.js (hunk 3 above)
Comment (jleft) [inline src/assets/js/chart/rsiChart.js]: If the issue is binding data appropriately, maybe it's multi (particularly the mapping functionality) to the rescue (again!)?

Something like this:

```js
var rsi = fc.indicator.renderer.relativeStrengthIndex();
var multi = fc.series.multi()
    .series([rsi])
    .mapping(function() {
            return data.data;
        }
    });
```

Comment (djmiley) [inline src/assets/js/chart/rsiChart.js]: Ok, yes, I was missing something. Thanks. :)

File: src/assets/js/chart/macdChart.js [hunk 7]
```diff
@@ -2,11 +2,23 @@
...
+
         var dispatch = d3.dispatch('viewChange');
 
-        var macdAlgorithm = fc.indicator.algorithm.macd();
+        var macdTimeSeries = fc.chart.linearTimeSeries()
+            .xAxisHeight(0)
+            .yAxisWidth(yAxisWidth)
+            .yOrient('right');
 
+        var zero = fc.annotation.line()
```
Comment (jleft) [inline src/assets/js/chart/macdChart.js]: No too fussed either way, but could just use a line?

File: src/assets/js/chart/macdChart.js [hunk 8]
```diff
@@ -14,33 +26,45 @@
...
+            var paddedYExtent = sc.util.domain.padYDomain([-maxYExtent, maxYExtent], 0.04);
+            macdTimeSeries.yDomain(paddedYExtent);
 
-            macd.xScale()
-                .domain(viewDomain)
-                .range([0, fc.util.innerDimensions(selection.node()).width]);
-            macd.yScale()
-                .domain([-maxYExtent, maxYExtent])
-                .range([fc.util.innerDimensions(selection.node()).height, 0]);
+            multi.mapping(function(series) {
```
Comment (jleft) [inline src/assets/js/chart/macdChart.js]: Can we set the mapping and decorate outside of this function, where multi is initialised?

File: src/assets/js/chart/rsiChart.js [hunk 9]
```diff
@@ -2,42 +2,60 @@
...
-            rsi.xScale()
-                .domain(viewDomain)
-                .range([0, fc.util.innerDimensions(selection.node()).width]);
-            rsi.yScale().range([fc.util.innerDimensions(selection.node()).height, 0]);
+            rsiTimeSeries.xDomain(viewDomain)
+                .yDomain([0, 100]);
 
             rsiAlgorithm(data);
 
+            multi.mapping(function(series) {
```
Comment (jleft) [inline src/assets/js/chart/rsiChart.js]: Similar comment to above - can we set the mapping outside of this function, where multi is initialised?

File: src/assets/js/chart/rsiChart.js [hunk 10]
```diff
@@ -2,42 +2,60 @@
...
+                    return data;
+                })
+                .decorate(function(g) {
+                    g.enter()
+                        .attr('class', function() {
+                            return 'multi';
```
Comment (jleft) [inline src/assets/js/chart/rsiChart.js]: Do we need to do this? I think this is the default behaviour?

File: src/assets/js/main.js [hunk 11]
```diff
@@ -28,7 +28,12 @@
         svgSecondary.datum(dataModel)
             .filter(function(d, i) { return i < secondaryChart.length; })
             .each(function(d, i) {
-                d3.select(this).call(secondaryChart[i]);
+                var selection = d3.select(this);
```
Comment (jleft) [inline src/assets/js/main.js]: How about this?

```js
d3.select(this)
    .attr('class', 'chart secondary ' + secondaryChart[i].valueString)
    .call(secondaryChart[i].option);
```

File: src/assets/styles/secondary.less [hunk 12]
```diff
@@ -1,3 +1,17 @@
-.secondary-row > .col-chart {
-  padding-right : calc(@bootstrap-padding-width + @y-axis-width);
-}
\ No newline at end of file
+.secondary {
+  .x-axis {
+
```
Comment (jleft) [inline src/assets/styles/secondary.less]: Remove line

File: package.json [hunk 13]
```diff
@@ -24,7 +24,7 @@
   "homepage": "https://github.com/scottlogic/d3fc-showcase#readme",
   "dependencies": {
     "bootstrap": "^3.3.5",
-    "d3fc": "^1.0.0"
+    "d3fc": "1.0.0"
```
Comment (jleft) [inline package.json]: Can we use 1.0.1?

File: src/assets/js/chart/rsiChart.js [hunk 14]
```diff
@@ -2,42 +2,53 @@
...
+        var yAxisWidth = 45;
+
         var dispatch = d3.dispatch('viewChange');
 
-        var rsiScale = d3.scale.linear()
-            .domain([0, 100]);
+        var rsi = fc.indicator.renderer.relativeStrengthIndex();
+        var multi = fc.series.multi()
+            .series([rsi])
+            .mapping(function(series) {
```
Comment (jleft) [inline src/assets/js/chart/rsiChart.js]: Don't need `series`.

File: src/assets/js/chart/macdChart.js (hunk 7 above)
Comment (djmiley) [inline src/assets/js/chart/macdChart.js]: Happy to remove it, but I think it makes the chart clearer.

Comment (jleft) [inline src/assets/js/chart/macdChart.js]: I was thinking that maybe we could just use a `line` element or `fc.series.line()` (or maybe even an axis without any ticks), as I don't think we need any of the labelling which the annotation offers, but it's probably not worth changing... happy to stick with the annotation.

I agree, I do think it looks a lot nicer with the line :)

File: package.json [hunk 15]
```diff
@@ -24,7 +24,7 @@
   "homepage": "https://github.com/scottlogic/d3fc-showcase#readme",
   "dependencies": {
     "bootstrap": "^3.3.5",
-    "d3fc": "^1.0.0"
+    "d3fc": "1.0.1"
```
Comment (jleft) [inline package.json]: Sorry... 1.2.0 might be more relevant now?

**Comment by jleft** [comment]
:+1: thanks @djmiley

**Pull request merged**